src/taipy/core/_entity/_submittable.py

```diff
@@ -37,10 +37,59 @@ def submit(
     ):
         raise NotImplementedError
 
-    def _get_inputs(self) -> Set[DataNode]:
-        dag = self._build_dag()
+    def __get_all_data_nodes_in_dag(self, dag: Optional[nx.DiGraph] = None) -> Set[DataNode]:
+        dag = self._build_dag() if not dag else dag
+        return {node for node in dag.nodes if isinstance(node, DataNode)}
+
+    def get_inputs(self, dag: Optional[nx.DiGraph] = None) -> Set[DataNode]:
+        """Return the set of input data nodes of the submittable entity.
+
+        Returns:
+            The set of input data nodes.
+        """
+        dag = self._build_dag() if not dag else dag
         return {node for node, degree in dict(dag.in_degree).items() if degree == 0 and isinstance(node, DataNode)}
 
+    def get_outputs(self, dag: Optional[nx.DiGraph] = None) -> Set[DataNode]:
+        """Return the set of output data nodes of the submittable entity.
+
+        Returns:
+            The set of output data nodes.
+        """
+        dag = self._build_dag() if not dag else dag
+        return {node for node, degree in dict(dag.out_degree).items() if degree == 0 and isinstance(node, DataNode)}
+
+    def get_intermediate(self) -> Set[DataNode]:
```

Comment:
```suggestion
    def get_intermediate(self, dag: Optional[nx.DiGraph] = None) -> Set[DataNode]:
```
Shouldn't we also accept a dag parameter here? Like get_inputs and get_output

Reply:
Since `get_intermediate` is exposed to the end user, I don't think we should accept an `nx.DiGraph` as a parameter.

Another way of doing it is:

``` python
def get_intermediate(self) -> Set[DataNode]:
    dag = self._build_dag()
    return {node for node, degree in dict(dag.degree).items() if degree != 0 and isinstance(node, DataNode)}
```

With this implementation, you simultaneously parse the dag and build the new set. You don't need three loops (`__get_all_data_nodes_in_dag`, `__get_inputs`, and `__get_outputs`).

Reply:
this will actually get all of the data nodes...

Reply:
if the degree (in + out) is != from 0 ?
Maybe the degree API does not exist or is not what I am thinking about then.

Reply:
never mind, it should be `return {node for node, degree in dict(dag.degree).items() if degree > 1 and isinstance(node, DataNode)}` :D